rpc/client/httpclient.go

```diff
@@ -85,7 +86,16 @@ var _ rpcClient = (*baseRPCClient)(nil)
 // NewHTTP takes a remote endpoint in the form <protocol>://<host>:<port> and
 // the websocket path (which always seems to be "/websocket")
 func NewHTTP(remote, wsEndpoint string) *HTTP {
-	rc := rpcclient.NewJSONRPCClient(remote)
+	httpClient := rpcclient.DefaultHTTPClient(remote)
+	return NewHTTPWithClient(remote, wsEndpoint, httpClient)
+}
+
+// NewHTTPWithClient allows for setting a custom http client. See NewHTTP
+func NewHTTPWithClient(remote, wsEndpoint string, client *http.Client) *HTTP {
```

Comment:
http.Client is meant to be passed around as a pointer and all methods are on the pointer version of the struct: https://golang.org/pkg/net/http/#Client
It would be weird to require it to be passed as a copy and might have unintended consequences.

Reply:
> It would be weird to require it to be passed as a copy and might have unintended consequences.

what consequences are u talking about?

Reply:
That would make copies of all properties in the struct. The struct is at: https://golang.org/src/net/http/client.go
It turns out that they all seem safe to copy because they are all pointers (interfaces). It's pretty unusual for someone to be modifying the http client after it's passed in.

Structs passed around as pointer versions of themselves is a pretty standard pattern, especially with stateful structs like long lived clients. That's why it's non-idomatic to dereference them, copy them and then pass them as arguments to another function. Every function that uses an http.Client would normally require the pointer version because all methods are on the pointer version.

For the caller to pass in the non-pointer, the caller would pretty much always have to just check for nil, dereference it and pass it in, which is not that different from checking for nil after it's passed in, but requires way more duplication of the check in all the callers.

If you still want to do this, I'll do it, but it feels wrong.

Reply:
That would make copies of all properties in the struct. The struct is at: https://golang.org/src/net/http/client.go
It turns out that they all seem safe to copy because they are all pointers (interfaces). It's pretty unusual for someone to be modifying the http client after it's passed in.

Structs passed around as pointer versions of themselves is a pretty standard pattern, especially with stateful structs like long lived clients. That's why it's non-idomatic to dereference them, copy them and then pass them as arguments to another function. Every function that uses an http.Client would normally require the pointer version because all methods are on the pointer version.

For the caller to pass in the non-pointer, the caller would pretty much always have to just check for nil, dereference it and pass it in, which is not that different from checking for nil after it's passed in, but requires way more duplication of the check in all the callers.

If you still want to do this, I'll do it, but it feels wrong.

Reply:
Right, DefaultClient returns a pointer. Okay, let's not do that. Thanks!